### PR TITLE
Fix carrier algorithm for copies with carriers

### PR DESCRIPTION
**Pull request by yetti**
http://ourweb.nla.gov.au/apps/jira/browse/DSCS-2100

This script updates original and master copies of OH&F with a carrier algorithm as based on the values in the lookup table for its carrier.

It's been ages since I've written Groovy, so there may be terribad things happening here. Please let me know if there's anything that needs to be burned with fire. Also, not too sure if the SQL is as efficient as it could be.

@scoen

### CODE REVIEW

File: carrierAlgorithm/src/main/groovy/au/gov/nla/CarrierAlgorithm.groovy [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+String PROP_AMBERDB_USERNAME = "amber.user";
+String PROP_AMBERDB_PASSWORD = "amber.pass";
+String PROP_AMBERDB_PATH = "amber.path";
+
+dbUrl = System.getProperty ( PROP_AMBERDB_URL )
+dbUser = System.getProperty ( PROP_AMBERDB_USERNAME )
+dbPassword = System.getProperty ( PROP_AMBERDB_PASSWORD )
+rootPath = System.getProperty ( PROP_AMBERDB_PATH )
+
+MysqlDataSource mds = new MysqlDataSource();
```
Comment (AaronCMuller) [inline carrierAlgorithm/src/main/groovy/au/gov/nla/CarrierAlgorithm.groovy]: Can prob just have 'def' but whatever.

File: carrierAlgorithm/src/main/groovy/au/gov/nla/CarrierAlgorithm.groovy [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    works.add(soundCopiesIterator.next())
+}
+
+handle.close()
+
+worksIterator = works.iterator()
+while (worksIterator.hasNext()) {
+    def next = worksIterator.next()
+    id = next.get("id")
+    amberSession = new AmberDb(ds, Paths.get(rootPath)).begin()
```
Comment (AaronCMuller) [inline carrierAlgorithm/src/main/groovy/au/gov/nla/CarrierAlgorithm.groovy]: I think this should be a few lines up, I don't think you want to create this in every iteration.

**Comment by AaronCMuller** [comment]
Looks good to me, although I didn't look at the SQL at all.

File: carrierAlgorithm/src/main/groovy/au/gov/nla/CarrierAlgorithm.groovy [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+def soundCopiesIterator = handle.createQuery(soundCopiesSql).iterator()
+def works = []
+while (soundCopiesIterator.hasNext()) {
+    works.add(soundCopiesIterator.next())
+}
+
+handle.close()
+
+worksIterator = works.iterator()
+amberSession = new AmberDb(ds, Paths.get(rootPath)).begin()
```
Comment (yetti) [inline carrierAlgorithm/src/main/groovy/au/gov/nla/CarrierAlgorithm.groovy]: @scoen Refactored to initialise `amberSession` outside of the `while-loop` and cleared the AmberGraph after calling `commit()`.

Merge?

**Pull request merged**